File: 349-intersection-of-two-arrays-ll.py
class Solution:
    def intersect(self, nums1: List[int], nums2: List[int]) -> List[int]:
        ans = []
        visited = {}

        for num in nums1:
            visited[num] = visited.get(num, 0) + 1
        for num in nums2:
            if num in visited and visited[num] > 0:
                visited[num] -= 1
                ans.append(num)
        return ans

File: 349-intersection-of-two-arrays-ll.java
class Solution {
    public int[] intersect(int[] nums1, int[] nums2) {
        Map<Integer,Integer> freq = new HashMap<>();
        List<Integer> ans =  new ArrayList<>();

        for(int num : nums1){
            freq.put(num,freq.getOrDefault(num,0)+1);
        }

        for(int num : nums2){
            if(freq.containsKey(num) && freq.getOrDefault(num,0)>0){
                freq.put(num,freq.get(num)-1);
                ans.add(num);
            }
        }
        int[] res = new int[ans.size()];
        int i = 0;

        for (int num : ans) {
            res[i++] = num;
        }

        return res;
    }
}